### PR TITLE
Athena 65

### LINKED ISSUE #65: Be able to tag a collection of vocabularies and share it with other collaborators
The new feature request (idea) is to allow a user tag a collection of vocabularies and share this tag/collection with other collaborators so that multiple institutions would be able to download the same  vocabularies.

@Karthik Natarajan

### PR DESCRIPTION
**Pull request by ssuvorov-fls**
fixes #65

### CODE REVIEW

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java [hunk 1]
```diff
@@ -206,11 +217,68 @@ public DownloadBundle saveDownloadItems(DownloadBundle bundle, List<Integer> idV
...
-        return converterUtils.convertList(history, DownloadBundleDTO.class);
+        List<DownloadBundle> history = downloadBundleRepository.findByUserId(user.getId(), sort);
+        Set<Long> bundleIds = history.stream()
+                .map(b -> b.getId())
+                .collect(Collectors.toSet());
+
+        List<DownloadShare> shares = downloadShareRepository.findByDownloadShareIdUserEmail(user.getEmail());
+        List<DownloadBundleDTO> sharedDTOs = new ArrayList<>();
+        // add shared bundles to list of available downloads
+        if (!shares.isEmpty()) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: How I can share download for the first time?
There is the following condition on the UI side:
```
const isShareable = isAlreadyShared || !bundle.downloadShareDTO;
```
And from server:
```
cdmVersion: 5
date: 1566830435154
downloadShareDTO: null
id: 3
link: null
name: "rxnorm"
releaseVersion: "v5.0 17-JUN-19"
status: "PENDING"
vocabularies: [{id: 8, code: "RxNorm", name: "", required: null, update: "2016-10-03", clickDefault: true,…}]
```

File: src/main/java/com/odysseusinc/athena/service/DownloadShareService.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.odysseusinc.athena.model.athena.DownloadBundle;
+import com.odysseusinc.athena.model.athena.DownloadShare;
+import com.odysseusinc.athena.model.security.AthenaUser;
+
+import java.util.List;
+
+public interface DownloadShareService {
+    List<DownloadShare> getBundleShares(Long downloadBundleId);
+
+    List<DownloadShare> getBundleShares(String shareUserEmail);
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/DownloadShareService.java]: Where this is used?

File: src/main/java/com/odysseusinc/athena/service/DownloadShareService.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    List<String> getUserEmails(Long downloadBundleId);
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/DownloadShareService.java]: Where this is used?

File: src/main/java/com/odysseusinc/athena/model/athena/DownloadShareId.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package com.odysseusinc.athena.model.athena;
+
+import org.hibernate.validator.constraints.NotBlank;
+
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+import javax.validation.constraints.NotNull;
+import java.io.Serializable;
+
+@Embeddable
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/model/athena/DownloadShareId.java]: Why composite key was used? What is the advantages of solution?

File: src/main/java/com/odysseusinc/athena/model/athena/DownloadShare.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "download_share")
+public class DownloadShare {
+    @EmbeddedId
+    private DownloadShareId downloadShareId;
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/model/athena/DownloadShare.java]: There is an implicit linkage to `DownloadBundle`. But why not to refer directly to it?
And no problems with orphans, no issue with search query.

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+@Service
+public class DownloadShareServiceImpl implements DownloadShareService {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: Does this not require to be transactional?

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        if (downloadShares == null || downloadShares.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return downloadShares.stream()
+                .map(bs -> bs.getUserEmail())
+                .collect(toList());
+    }
+
+    @Override
+    public List<DownloadShare> change(DownloadBundle bundle, DownloadShareChangeDTO changeDTO, AthenaUser user) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: Oh, mix of DTOs and entities looks like design mistake.

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        if (changeDTO == null) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: And could you simplify this? About 40 lines of code just to update collections of emails.

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java [hunk 9]
```diff
@@ -206,11 +217,68 @@ public DownloadBundle saveDownloadItems(DownloadBundle bundle, List<Integer> idV
...
+            for(DownloadShare share: shares) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: There are lot of queries and conversions to get list of shares.

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java [hunk 10]
```diff
@@ -206,11 +217,68 @@ public DownloadBundle saveDownloadItems(DownloadBundle bundle, List<Integer> idV
...
+                }
+            }
+        }
+
+        List<DownloadBundleDTO> dtos = converterUtils.convertList(history, DownloadBundleDTO.class);
+
+        // Get list of shared bundles where owner is current user
+        List<DownloadShare> ownerShares = downloadShareRepository.findByOwnerId(user.getId());
+        dtos.stream()
+                .forEach(dto -> {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: `forEach` does not require `stream()`

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java [hunk 11]
```diff
@@ -206,11 +217,68 @@ public DownloadBundle saveDownloadItems(DownloadBundle bundle, List<Integer> idV
...
+                    String emails = ownerShares.stream()
+                            .filter(o -> o.getBundleId() == dto.getId())
+                            .map(o -> o.getUserEmail())
+                            .collect(Collectors.joining(", "));
+                    // if we get list of users with whom this bundle was shared
+                    // then the current user is the owner of this bundle
+                    if (emails != null && !emails.isEmpty()) {
+                        DownloadShareDTO ownerDto = new DownloadShareDTO();
```
Comment (wivern) [inline src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: Anyway why not converter?

File: src/main/java/com/odysseusinc/athena/service/DownloadShareService.java (hunk 2 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/DownloadShareService.java]: fixed

File: src/main/java/com/odysseusinc/athena/service/DownloadShareService.java (hunk 3 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/DownloadShareService.java]: fixed

File: src/main/java/com/odysseusinc/athena/model/athena/DownloadShare.java (hunk 5 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/model/athena/DownloadShare.java]: fixed

File: src/main/java/com/odysseusinc/athena/model/athena/DownloadShareId.java (hunk 4 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/model/athena/DownloadShareId.java]: fixed

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java (hunk 6 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: fixed

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java (hunk 7 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: fixed

File: src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java (hunk 8 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/DownloadShareServiceImpl.java]: refactoring made

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java (hunk 9 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: refactoring made

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java (hunk 10 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: fixed

File: src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java (hunk 11 above)
Comment (ssuvorov-fls) [reply src/main/java/com/odysseusinc/athena/service/impl/VocabularyServiceImpl.java]: refactoring made

**Pull request merged**